monitoring/storage.go

```diff
@@ -0,0 +1,238 @@
+	err := c.evalPath()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.NewAggregate(c.checkFsType(ctx, reporter),
+		c.checkCapacity(ctx, reporter),
+		c.checkWriteSpeed(ctx, reporter))
+}
+
+// cleanPath returns fully evaluated path with symlinks resolved
+// if path doesn't exist but should be created, then it returns
+// first available parent directory, and checks will be applied to it
+func (c *StorageChecker) evalPath() error {
+	fi, err := os.Stat(c.Path)
+
+	if os.IsNotExist(err) && !c.WillBeCreated {
+		return trace.BadParameter("%s does not exist", c.Path)
+	}
+
+	p := c.Path
+	for {
+		fi, err = os.Stat(p)
+		if err == nil && fi.IsDir() {
+			c.path = p
+			return nil
+		} else if err == nil && !fi.IsDir() {
+			return trace.BadParameter("%s is not a directory", p)
+		}
+
+		parent := filepath.Dir(p)
+		if parent == p {
+			// shouldn't happen: root reached and it's not a dir
+			return trace.BadParameter("%s is root and is not a dir", p)
+		}
+		p = parent
+	}
+}
+
+func (c *StorageChecker) checkFsType(ctx context.Context, reporter health.Reporter) error {
+	if len(c.Filesystems) == 0 {
+		return nil
+	}
+
+	mnt, err := fsFromPath(c.path)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	probe := &pb.Probe{Checker: c.Name()}
+
+	if utils.StringInSlice(c.Filesystems, mnt.SysTypeName) {
+		probe.Status = pb.Probe_Running
+	} else {
+		probe.Status = pb.Probe_Failed
+		probe.Detail = fmt.Sprintf("path %s requires %v, belongs to %s mount point of type %s",
+			c.Path, c.Filesystems, mnt.DirName, mnt.SysTypeName)
+	}
+	reporter.Add(probe)
+	return nil
+}
+
+type childPathFirst []sigar.FileSystem
+
+func (a childPathFirst) Len() int           { return len(a) }
+func (a childPathFirst) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a childPathFirst) Less(i, j int) bool { return strings.HasPrefix(a[i].DirName, a[j].DirName) }
+
+func fsFromPath(path string) (*sigar.FileSystem, error) {
+	cleanpath, err := filepath.EvalSymlinks(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	mounts := sigar.FileSystemList{}
+	err = mounts.Get()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sort.Sort(childPathFirst(mounts.List))
+
+	for _, mnt := range mounts.List {
+		if strings.HasPrefix(cleanpath, mnt.DirName) {
+			return &mnt, nil
+		}
+	}
+
+	return nil, trace.BadParameter("failed to locate filesystem for %s", path)
+}
+
+func (c *StorageChecker) checkCapacity(ctx context.Context, reporter health.Reporter) error {
+	var stat syscall.Statfs_t
+
+	err := syscall.Statfs(c.path, &stat)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	avail := uint64(stat.Bsize) * stat.Bavail
+	if avail < c.MinFreeBytes {
+		reporter.Add(&pb.Probe{
+			Checker: c.Name(),
+			Detail: fmt.Sprintf("%s available space left on %s, minimum of %s required",
+				humanize.Bytes(avail), c.Path, humanize.Bytes(c.MinFreeBytes)),
+			Status: pb.Probe_Failed,
+		})
+	}
+
+	return nil
```

Comment:
Same as below - maybe add a successful probe to indicate the check passed like we do for other checks?

Reply:
it's there, see line 75